test_generators/operations/deposits.py

```diff
@@ -0,0 +1,168 @@
+        message_hash=signing_root(deposit_data),
+        privkey=privkey,
+        domain=spec.get_domain(
+            state.fork,
+            spec.get_current_epoch(state),
+            spec.DOMAIN_DEPOSIT,
+        )
+    )
+    return deposit_data
+
+
+def build_deposit(state,
+                  deposit_data_leaves: List[spec.Bytes32],
+                  pubkey: spec.BLSPubkey,
+                  withdrawal_cred: spec.Bytes32,
+                  privkey: int,
+                  amount: int) -> spec.Deposit:
+
+    deposit_data = build_deposit_data(state, pubkey, withdrawal_cred, privkey, amount)
+
+    item = spec.hash(deposit_data.serialize())
+    index = len(deposit_data_leaves)
+    deposit_data_leaves.append(item)
+    tree = calc_merkle_tree_from_leaves(tuple(deposit_data_leaves))
+    proof = list(get_merkle_proof(tree, item_index=index))
+
+    deposit = spec.Deposit(
+        proof=list(proof),
+        index=index,
+        data=deposit_data,
+    )
+    assert spec.verify_merkle_branch(item, proof, spec.DEPOSIT_CONTRACT_TREE_DEPTH, index, get_merkle_root(tuple(deposit_data_leaves)))
+
+    return deposit
+
+
+def build_deposit_for_index(initial_validator_count: int, index: int) -> Tuple[spec.Deposit, spec.BeaconState]:
+    genesis_deposits = genesis.create_deposits(
+        keys.pubkeys[:initial_validator_count],
+        keys.withdrawal_creds[:initial_validator_count]
+    )
+    state = genesis.create_genesis_state(genesis_deposits)
+
+    deposit_data_leaves = [spec.hash(dep.data.serialize()) for dep in genesis_deposits]
+
+    deposit = build_deposit(
+        state,
+        deposit_data_leaves,
+        keys.pubkeys[index],
+        keys.withdrawal_creds[index],
+        keys.privkeys[index],
+        spec.MAX_DEPOSIT_AMOUNT,
+    )
+
+    state.latest_eth1_data.deposit_root = get_merkle_root(tuple(deposit_data_leaves))
+    state.latest_eth1_data.deposit_count = len(deposit_data_leaves)
+
+    return deposit, state
+
+
+@to_dict
+def valid_deposit():
+    new_dep, state = build_deposit_for_index(10, 10)
+    yield 'description', 'valid deposit to add new validator'
+    yield 'pre', encode(state, spec.BeaconState)
+    yield 'deposit', encode(new_dep, spec.Deposit)
+    spec.process_deposit(state, new_dep)
+    yield 'post', encode(state, spec.BeaconState)
+
+
+@to_dict
+def valid_topup():
+    new_dep, state = build_deposit_for_index(10, 3)
+    yield 'description', 'valid deposit to top-up existing validator'
+    yield 'pre', encode(state, spec.BeaconState)
+    yield 'deposit', encode(new_dep, spec.Deposit)
+    spec.process_deposit(state, new_dep)
+    yield 'post', encode(state, spec.BeaconState)
+
+
+@to_dict
+def invalid_deposit_index():
+    new_dep, state = build_deposit_for_index(10, 10)
+    # Mess up deposit index, 1 too small
+    state.deposit_index = 9
+
+    yield 'description', 'invalid deposit index'
+    yield 'pre', encode(state, spec.BeaconState)
+    yield 'deposit', encode(new_dep, spec.Deposit)
```

Comment:
Feels a bit weird to not run it to ensure `process_deposit` throws..
That said, it's not a test.

Reply:
Hmm, yes, here the border between pytests and test-generation is fuzzy. I do think we can make assumptions in the generators, as they are not tests themselves. Happy to discuss possible ways to merge in py-tests in a clean way, and get these types of guarantees when generating tests.

Reply:
Ok, looks like a sanity assertion is appropriate here to prevent future code from unknowingly breaking the test generation. Still hope to see something in place to reduce py-tests into formal definitions of expectations/invariants, and then just check these during test-generation. This way we remove the need for any random data generation in the invariants definition (effectively the py-tests now), and apply it to the widest variation of data available, aside from fuzzing: the full eth2.0 test suite. And we could re-use these invariant-checkers for fuzzing :)